[Tune] Sync trial artifacts to cloud (#32334)
ray-project/ray: doc/source/tune/tutorials/tune-output.rst

Logging and Outputs in Tune
===========================

> Tune by default will log results for TensorBoard, CSV, and JSON formats.

Reviewer: ```suggestion
By default, Tune logs results for TensorBoard, CSV, and JSON formats.
```

.. _trainable-logging:

How do you log arbitrary files from a Tune Trainable?
-----------------------------------------------------

By default, Tune only logs the *training result dictionaries* and *checkpoints* from your Trainable.

> However, you may want to save a file that visualizes the model weights/model graph,

Reviewer: ```suggestion
However, you may want to save a file that visualizes the model weights or model graph,
```

Reviewer: I hope that doesn't change the meaning of what you want to say. In general, we try to avoid using slashes because they are ambiguous.

            for step in range(100):
                library.log_model(...)
                library.log(results, step=step)

                # You can also just write to a file directly.

>                 with open(f"./artifact_{step}.txt", "w") as f:

Reviewer 2: can you note something about the local dir being a trial directory?

Author: I have a few sentences about this right below, but will also add a bit of info to the inline comment.

                    f.write("Artifact Data")

                session.report(results)

.. tabbed:: Class API

    .. code-block:: python

>         import library

Reviewer 2: rename to logging_library? or mlflow?

                    id=trial_id,
                    resume=trial_id,
                    reinit=True,
                    allow_val_change=True
                )
                library.set_log_path(os.getcwd())

            def step(self):
                library.log_model(...)

>                 # You can also just write to a file directly.

Reviewer: ```suggestion
                # You can also write to a file directly.
```

                step = result["training_iteration"]
                library.log(res_dict, step=step)

> In the code snippet above, ``library`` refers to whatever 3rd party logging library you are using.

Reviewer: ```suggestion
In the code snippet above, ``library`` refers to the third party logging library you are using.
```

Note that ``library.set_log_path(os.getcwd())`` is an imaginary API that we are using
for demonstation purposes, and it highlights that the third-party library
should be configured to log to the Trainable's *working directory.* By default,

> the current working directory of both functional and class trainables will be set to the

Reviewer: ```suggestion
the current working directory of both functional and class trainables are set to the
```

> the current working directory of both functional and class trainables will be set to the

Reviewer: ```suggestion
the current working directory of both functional and class trainables is set to the
```

corresponding trial directory once it's been launched as a remote Ray actor.

When running with multiple nodes using the :ref:`default syncing method <tune-default-syncing>`,

> trial artifacts will be synchronized to the driver node under the specified path.

Reviewer: ```suggestion
trial artifacts are synchronized to the driver node under the specified path.
```

This will allow you to visualize and analyze logs of all distributed training workers on a single machine.

> When :ref:`specifying a cloud upload directory <tune-cloud-checkpointing>`, trial artifacts will be uploaded to that cloud bucket

Reviewer: ```suggestion
When :ref:`specifying a cloud upload directory <tune-cloud-checkpointing>`, trial artifacts are uploaded to that cloud bucket
```

> for later analysis. Note that in this case, the driver node will not necessarily contain

Reviewer: ```suggestion
for later analysis. Note that the driver node does not necessarily contain
```

artifacts from *all* trials -- only the ones that were running on that node.
To disable artifacts from being uploaded to the cloud, set ``SyncConfig(sync_artifacts=False)`` in :class:`~ray.tune.syncer.SyncConfig`.

.. warning::

>     Appending to trial artifacts upon restoration is not currently supported.

Reviewer: ```suggestion
    Appending to trial artifacts upon restoration is not supported.
```

>     To get around this, we recommend saving trial artifacts a separate files with unique filenames.

Reviewer: ```suggestion
    As a workaround, save trial artifacts to separate files with unique filenames.
```

    For example, instead of doing this:

    .. code-block:: python

        def appending_train_fn(config):
            for i in range(config["num_epochs"]):
                with open("./artifact.txt", "a") as f:
                    f.write(f"Some data about iteration {i}\n")

> 

Reviewer 2: We don't want to recommend this generically? I think you want to say specifically  that you will want to do this only when you need to append, but for example if the file is to be rewritten (like a checkpoint) it's probably ok right?

Author: Actually, even if it's rewritten, as long as the filename is the same, it'll keep getting overwritten by the driver's old copy of the artifact. So, unique files are kind of needed for restore behavior to work.

>     You should log artifacts as independent files with unique filenames:

Reviewer: ```suggestion
    Log artifacts as independent files with unique filenames:
```

    .. code-block:: python

        def separate_files_train_fn(config):
            for i in range(config["num_epochs"]):
                with open(f"./artifact_{i}.txt", "w") as f:
                    f.write(f"Some data about iteration {i}\n")

>     If you are running into issues with this, please `file an issue <https://github.com/ray-project/ray/issues>`_!

Reviewer: ```suggestion
    If you are running into issues, `file an issue <https://github.com/ray-project/ray/issues>`_!
```

Reviewer: I am not familiar with the syntax of the underscore at the end of the line. I wanted to suggest that you remove the exclamation point.
